src/main/resources/META-INF/rewrite/micronaut3-to-4.yml

```diff
@@ -0,0 +1,376 @@
+      newVersion: 4.0.0-M2
+  - org.openrewrite.java.micronaut.UpgradeMicronautGradlePropertiesVersion:
+      newVersion: 4.0.0-M2
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateBuildPlugins
+displayName: Add Micronaut build plugins to 4.x
+description: This recipe will update the shadow jar plugin to 8.x and the Micronaut build plugins to 4.x for a Gradle build.
+recipeList:
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: com.github.johnrengelman.shadow
+      newVersion: 8.x
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.aot
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.application
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.crac
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.docker
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.graalvm
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.library
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.minimal.application
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.minimal.library
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.test-resources
+      newVersion: 4.0.0-M2
+  - org.openrewrite.gradle.plugins.UpgradePluginVersion:
+      pluginIdPattern: io.micronaut.test-resources-consumer
+      newVersion: 4.0.0-M2
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateMicronautPlatformBom
+displayName: Update to Micronaut 4.x platform BOM
+description: This recipe will update a Gradle or Maven build to reference the Micronaut 4 platform BOM.
+recipeList:
+  - org.openrewrite.maven.ChangeParentPom:
+      oldGroupId: io.micronaut
+      oldArtifactId: micronaut-parent
+      newGroupId: io.micronaut.platform
+      newVersion: 4.0.0-M2
+      retainVersions:
+        - io.micronaut:micronaut-validation
+        - io.micronaut:micronaut-session
+---
+type: specs.openrewrite.org/v1beta/recipe
+name: org.openrewrite.java.micronaut.UpdateJakartaAnnotations
+displayName: Update jakarta annotations dependency
+description: This recipe will remove jakarta annotations dependency as it is a transitive dependency of micronaut-inject, and migrate from javax.annotation if needed.
+recipeList:
+  - org.openrewrite.java.ChangePackage:
+      oldPackageName: javax.annotation
+      newPackageName: jakarta.annotation
+      recursive: true
```

Comment:
For these types of changes I would have expected the reuse of the matching recipe n rewrite-migrate-java:
https://github.com/openrewrite/rewrite-migrate-java/blob/1e8a08b91b66f3c3f8e7fcd9ed9f6d772aa4083f/src/main/resources/META-INF/rewrite/jakarta-ee-9.yml#L92-L114
That would then come down to something like this here:
```yaml
 - org.openrewrite.java.migrate.jakarta.JavaxMigrationToJakarta
```


What made you chose to duplicate some of those existing recipes here?

Reply:
I think I was just trying to be more targeted, because we already had users migrated to jakarta in many cases in Micronaut 3. Also, note that we are actually removing the jakarta.annotation-api dependency in this specific case because it now gets pulled in transitively in Micronaut 4. That said, I think I missed the existence of `org.openrewrite.java.migrate.jakarta.JavaxAnnotationPackageToJakarta` which does what we need in terms of changing package usage. I will swap that in here.